[llvm] Improve implementation of StringRef::find_last_of and cie

serge-sans-paille opened the pull request:
Using a simple LUT implies less operations than a bitset, at the expense of slightly more stack usage.

Mandatory performance report:

https://llvm-compile-time-tracker.com/compare.php?from=3f9d385e5844f2f1f144305037cfc904789c6187&to=86f685e9552acaf8d5db630161e91752348422a1&stat=instructions:u

aganea commented:
Hello @serge-sans-paille! Could you please explain a bit more in detail why do you think this is a good optimization? Less instructions retired is not necessary better if that increases the memory pressure? It might be nothing, but before the whole bitset was fitting in a cacheline, now it takes 4 cache lines to do the same thing. Again, this might be nothing, but that means 3 extra cache line being evicted elsewhere to execute this function. Also I'd like to be convinced by the profile link that you've attached, but the difference is so small that is could be normal fluctuations (also 'cycles' stat is higher). One more point, the previous bitshifting code in bitset might run better on AArch64 than on x64, so it'd be interesting to see the difference there too. Thanks!

serge-sans-paille commented:
@aganea I submitted a different approach that will work with any SSE2 - powered machine, and it doesn't have the same stack usage issue as the previous one.

kazutakahirata reviewed (commented):
This patch looks good to me.  I am not sure if we want to implement architecture-specific optimizations like `__SSE2__` at the source-code level.  But then I don't want that to block the landing of this patch.  Is there any way you could separate these two patches?  Thanks!

EDIT: I guess "this patch" isn't clear.  I am referring to the first one that replaces `"."` with `'.'`, etc.

joker-eph commented on llvm/lib/Support/StringRef.cpp:
```diff
@@ -274,6 +278,23 @@ StringRef::size_type StringRef::find_first_not_of(StringRef Chars,
+    } while (Sz);
+    return npos;
+  }
+#endif
```
Can this be abstracted or made out-of-line?
I'm wondering about the scalability of HW-specific intrinsics in-line (anticipating for the incoming `#elif defined(ARM64)`...)

serge-sans-paille replied:
I don't think using https://github.com/xtensor-stack/xsimd is an option :-) And https://en.cppreference.com/w/cpp/experimental/simd/simd is still not a thing :-/

We already have some bits of SSE2 in clang and llvm. OK to factor this in a function.

joker-eph replied:
Yeah I'm fine with SSE2, I was just trying to keep the specialized implementation out-of-line, what you have now looks good I think.

joker-eph commented on llvm/lib/Support/StringRef.cpp:
```diff
@@ -268,17 +272,44 @@ StringRef::size_type StringRef::find_first_not_of(StringRef Chars,
 
+#ifdef __SSE2__
+
+StringRef::size_type vectorized_find_last_of_specialized(const char *Data,
+                                                         size_t Sz, char C0,
```
static?

joker-eph replied:
```suggestion
static StringRef::size_type vectorized_find_last_of_specialized(const char *Data,
                                                         size_t Sz, char C0,
```

aganea commented on llvm/lib/Support/StringRef.cpp:
```diff
@@ -268,17 +272,44 @@ StringRef::size_type StringRef::find_first_not_of(StringRef Chars,
+  __m128i Needle1 = _mm_set1_epi8(C1);
+  do {
+    Sz = Sz < 16 ? 0 : Sz - 16;
+    __m128i Buffer = _mm_loadu_si128((const __m128i *)(Data + Sz));
```
This load instruction will generate an out-of-bounds access if `strlen(Data)<15`.  Do you think you can use the "slow" algorithm when that happens?

aganea commented on llvm/lib/Support/StringRef.cpp:
```diff
@@ -623,8 +657,7 @@ hash_code llvm::hash_value(StringRef S) {
 unsigned DenseMapInfo<StringRef, void>::getHashValue(StringRef Val) {
-  assert(Val.data() != getEmptyKey().data() &&
-         "Cannot hash the empty key!");
+  assert(Val.data() != getEmptyKey().data() && "Cannot hash the empty key!");
```
Can you please commit all these NFC formatting changes separately?

aganea commented on llvm/lib/Support/StringRef.cpp:
```diff
@@ -372,7 +402,8 @@ size_t StringRef::count(StringRef Str) const {
   size_t N = Str.size();
-  // TODO: For an empty `Str` we return 0 for legacy reasons. Consider changing
+  // TODO: For an empty `Str` we return 0 for legacy reasons. Consider
+  // changing
```
Text should go with the following line.

serge-sans-paille committed:
[llvm] Improve implementation of StringRef::find_last_of for the usual case of 2 chars

Almost all usage of StringRef::find_last_of in Clang/LLVM use a Needle
of 2 elements, which can be optimized using a generic vectorized
algorithm and a few bit hacks.

serge-sans-paille commented:
@aganea / @joker-eph / @kazutakahirata : I actually switched the implementation to something that's almost as efficient (no vector load required) and works across all architecture. (Not to mention it also increases my geekness karma)

aganea commented on llvm/lib/Support/StringRef.cpp:
```diff
@@ -268,17 +268,47 @@ StringRef::size_type StringRef::find_first_not_of(StringRef Chars,
                                              size_t From) const {
+  size_type Sz = std::min(From, Length);
+
+  if (Chars.size() == 2)
```
Could you please elaborate a bit more on the specific use cases that you hit in _clangd_ that justify this `if (Chars.size() == 2)`? At least a unit test along with a little comment about the "why" will help future readers of this patch.

serge-sans-paille commented on llvm/lib/Support/StringRef.cpp:
```diff
@@ -268,17 +268,47 @@ StringRef::size_type StringRef::find_first_not_of(StringRef Chars,
+/// This is a hot spot for some clangd operations, enough to be eligible to
+/// a vectorized implementation.
+static StringRef::size_type
+vectorized_find_last_of_specialized(const char *Data, size_t Sz, char C0,
```
I think yes, because it is indeed a vectorized version, but without SIMD instructions :-)

serge-sans-paille commented on llvm/lib/Support/StringRef.cpp:
```diff
@@ -268,17 +268,47 @@ StringRef::size_type StringRef::find_first_not_of(StringRef Chars,
   for (char C : Chars)
     CharBits.set((unsigned char)C);
 
-  for (size_type i = std::min(From, Length) - 1, e = -1; i != e; --i)
```
well, there's a call to `std::min` in the function prologue, this correctly defines the behavior of the function.

aganea replied:
Ah yes, thanks!

serge-sans-paille commented:
It turns out `StringRef::find_last_of` was missing unittests, I added that.

aganea commented on llvm/lib/Support/StringRef.cpp:
```diff
@@ -268,17 +268,47 @@ StringRef::size_type StringRef::find_first_not_of(StringRef Chars,
+  do {
+    Sz = Sz < 8 ? 0 : Sz - 8;
+    uint64_t Buffer = 0;
+    std::memcpy((void *)&Buffer, (void *)(Data + Sz), sizeof(Buffer));
```
Thanks for changing this!

aganea commented on llvm/lib/Support/StringRef.cpp:
```diff
@@ -268,17 +268,70 @@ StringRef::size_type StringRef::find_first_not_of(StringRef Chars,
                                              size_t From) const {
+  size_type Sz = std::min(From, Length);
+
+  if (Chars.size() == 2) {
```
Before going further with this PR, I'd like to fully understand which specific call site of `find_last_of` in ClangD is improved by this change, and what is the user flow that triggers it? Is this for `.find_last_of("\r\n")` ? Or for `.find_last_of("/\\");` ? Something else? The situation might change with different/newer CPU architectures, that's why a specific repro would be good to add in the PR description.

You also said above that this function appeared on a teammates' profile during code indexing, are you able to test before/after this patch see how the situation is improved? It would be really nice to have numbers along with this PR, so that others can compare.

aganea replied:
Is there any precedent for such micro-optimization in core libraries? I still feel that this optimization should go in a ClangD utility file instead. @dwblaikie @MaskRay @joker-eph Do you have an opinion on this?

joker-eph replied:
It's a tradeoff between impact vs cost of maintenance.

Moving it to clangd does not reduce the maintenance aspect for the LLVM project though, are we concerned with StringRef maintenance here?
I would say that on the contrary, moving it to clangs runs the risk than Bolt or LLDB reimplement the same thing, not knowing clangd has it.
So I'd say either the implementation isn't overly complex for the long term maintenance here compared to the benefits, or it may not belong to the project?

I don't find a mention of the current perf impact on clangd though?
In https://github.com/llvm/llvm-project/pull/71865#issuecomment-1981763725 you mention a micro-benchmark @serge-sans-paille but no the result as far as I can see?

arsenm commented:
Is this still relevant?

serge-sans-paille commented:
yeah, it probably is,I just switched priorities, I'll do another iteration on that one.

serge-sans-paille commented:
Code updated, and I ran the following micro benchmark:

```
$ cat a.cpp
#include "llvm/ADT/StringRef.h"
#include <cstdlib>

int main(int argc, char ** argv) {
  long n = atol(argv[2]);
  char * data = argv[1];
  llvm::StringRef S(data);
  size_t k = 0;
  for (long i = 0; i < n; ++i)
    k+= S.find_last_of("\r\n");
  return k;
}
$ clang++ a.cpp -fPIC -o a -O2 -lLLVM -L$PWD/lib
$ LD_LIBRARY_PATH=$PWD/lib ./a "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA" 10000000
```

which is as stupid as a micro benchmark can be. This patch make it run twice as fast as the non-patched version.

serge-sans-paille commented:
the build issue seems unrelated (?)

arsenm commented:
> the build issue seems unrelated (?)

The windows bot has been pure noise for many weeks

joker-eph commented:
> > the build issue seems unrelated (?)
> 
> The windows bot has been pure noise for many weeks

This is factually incorrect: https://lab.llvm.org/buildbot/#/builders/271
Your experience is likely outdated.

aganea commented on llvm/lib/Support/StringRef.cpp:
```diff
@@ -268,17 +268,47 @@ StringRef::size_type StringRef::find_first_not_of(StringRef Chars,
+  return ((v)-0x0101010101010101UL) & ~(v) & 0x8080808080808080UL;
+}
+static inline uint64_t hasvalue(uint64_t x, char n) {
+  return haszero((x) ^ (~0UL / 255 * (n)));
```
This hasn't been addressed yet.

dwblaikie commented:
Does this micro-optimization adversely affect cases that aren't covered by it (eg: does checking for length 2 slow down non-length-2 cases perceptibly?)

Ralender commented on llvm/lib/Support/StringRef.cpp:
```diff
@@ -268,17 +268,70 @@ StringRef::size_type StringRef::find_first_not_of(StringRef Chars,
+           0x7F7F7F7F7F7F7F7FULL);
+}
+static inline uint64_t hasvalue(uint64_t x, char n) {
+  return haszero((x) ^ (~0ULL / 255 * (n)));
```
Maybe its just me but I think `haszero((x) ^ ((uint64_t)n * 0x0101010101010101))` makes it more clear that `n` is just repeated over the entire `uint64_t`

Ralender commented on llvm/lib/Support/StringRef.cpp:
```diff
@@ -268,17 +268,70 @@ StringRef::size_type StringRef::find_first_not_of(StringRef Chars,
+  while (Sz >= 8) {
+    Sz -= 8;
+    uint64_t Buffer = 0;
+    std::memcpy((void *)&Buffer, (void *)(Data + Sz), sizeof(Buffer));
```
Isn't there some already existing utility to do unaligned loads more explicitly ?
